include/graphblas/algorithms/amg/amg_data.hpp

```diff
@@ -0,0 +1,201 @@
+				system_size( sys_size ), A( sys_size, sys_size ), A_diagonal( sys_size ), z( sys_size ), r( sys_size ),
+				smoother_temp( sys_size ) {}
+			// for safety, disable copy semantics
+			system_data( const system_data & o ) = delete;
+			system_data & operator=( const system_data & ) = delete;
+		};
+
+		/**
+		 * Data container for all multi-grid inputs and outputs.
+		 *
+		 * @tparam IOType Type of values of the vectors for intermediate results
+		 * @tparam NonzeroType Type of the values stored inside the system matrix \p A
+		 *                     and the coarsening matrix #Ax_finer
+		 *
+		 * This data structure stores information for a full multi-grid V cycle, i.e.
+		 * - input and output vectors for solution, residual and temporary vectors
+		 * - coarsening information, in particular the #coarsening_matrix that
+		 *   coarsens a larger system of size #finer_size to the current system
+		 *   of size #system_size
+		 * - the next level of coarsening, pointed to by #coarser_level,
+		 *        possibly being \c nullptr if no further coarsening is desired;
+		 *        note that this information is automatically
+		 *        destructed on object destruction (if any)
+		 *
+		 * Vectors stored here refer to the \b coarsened system
+		 * (with the exception of #Ax_finer),
+		 * thus having size #system_size; this also holds for the system matrix #A,
+		 * while #coarsening_matrix has size #system_size \f$ \times \f$ #finer_size.
+		 * Hence, the typical usage of this data structure is to coarsen \b external
+		 * vectors, e.g. vectors coming from another
+		 * \code multi_grid_data<IOType, NonZeroType>
+		 * \endcode object whose #system_size equals
+		 * \code this-> \endcode #fines_size, via
+		 * \code this-> \endcode #coarsening_matrix and store the coarsened
+		 * vectors internally. Mimicing the recursive behavior of standard multi-grid
+		 * simulations, the information for a further coarsening is stored inside
+		 * #coarser_level, so that the  hierarchy of coarsened levels is reflected
+		 * inside this data structure.
+		 *
+		 * As for \ref system_data, internal vectors and matrices are initialized to the
+		 * proper size, but their values are \b not initialized.
+		 */
+		template< typename IOType, typename NonzeroType >
+		struct multi_grid_data : public system_data< IOType, NonzeroType > {
+
+			const std::size_t finer_size; ///< ssize of the finer system to coarse from;
+			grb::Vector< IOType > Ax_finer; ///< finer vector for intermediate computations, of size #finer_size
+			grb::Matrix< NonzeroType > coarsening_matrix; ///< matrix of size #system_size \f$ \times \f$ #finer_size
+			                                              ///< to coarsen an input vector of size #finer_size
+			                                              ///< into a vector of size #system_size
+			struct multi_grid_data< IOType, NonzeroType > *coarser_level; ///< pointer to next coarsening level, for recursive
```

Comment:
Code style: multi_grid_data -> MultiGridData

C vs. C++: why not have the MultiGridData as an std::vector of structs that contain finer_size, Ax_finer, and coarsening_matrix ?